code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraint.java

```diff
@@ -0,0 +1,96 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DiskSizeConstraint.class);
+
+    private Long reserveSize;
+    private String volumeName;
+    private ObjectManager objectManager;
+
+    public DiskSizeConstraint(String volumeName, String sizeWithUnit, ObjectManager objMgr) {
+        this.volumeName = volumeName;
+        this.reserveSize = Long.parseLong(sizeWithUnit.replaceAll("[^0-9]", ""));
+        this.objectManager = objMgr;
+    }
+
+    @Override
+    public boolean matches(AllocationAttempt attempt, AllocationCandidate candidate) {
+        Set<Long> hostIds = candidate.getHosts();
+
+        // if one of the host does not have enough free space then return false
+        for (Long hostId : hostIds) {
+            boolean oneGood = false;
+
+            CacheManager cm = CacheManager.getCacheManagerInstance(this.objectManager);
+            HostInfo hostInfo = cm.getHostInfo(hostId);
+
+            // we will get a bunch of disks for that host and we need at least one disk with
+            // free space large enough
+            oneGood = checkDiskSize(cm, hostInfo, attempt.getInstanceId());
+
+            // if no disk with big enough free space for this host, then
+            // candidate is no good
+            if (!oneGood) {
+                log.debug("Scheduling instance [{}] to host [{}] rejected", attempt.getInstanceId(), hostId);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public boolean checkDiskSize (CacheManager cm, HostInfo hostInfo, Long instanceId) {
+
+        for (Entry<String, DiskInfo> entry : hostInfo.getAllDiskInfo()) {
+            DiskInfo diskInfo = entry.getValue();
+            Long capacity = entry.getValue().getCapacity();
+            Long allocatedSize = ((DiskInfo) diskInfo).getAllocatedSize();
+            Long freeSize = capacity - allocatedSize;
+
+            InstanceInfo instanceInfo = cm.getInstanceInfoForHost(hostInfo.getHostId(), instanceId);
+            
+            // we need to consider more than one volume constraints for this
+            // container case. That means free size will subtract all reserved
+            // size by other volumes as well beside allocated size.
+            InstanceDiskReserveInfo diskReserved = instanceInfo.getDiskReserveInfo(((DiskInfo)diskInfo).getDiskDevicePath());
+            if (diskReserved != null) {
+                freeSize -= diskReserved.getReservedSize();
+            }
+            if (freeSize >= this.reserveSize) {
+                // reserve disk for this instance
+                if (diskReserved != null) {
+                    diskReserved.addReservedSize(this.reserveSize);
+                } else {
+                    diskReserved = new InstanceDiskReserveInfo(((DiskInfo)diskInfo).getDiskDevicePath(), this.reserveSize, this.volumeName);
+                    instanceInfo.reserveDisk(diskReserved);
```

Comment:
I have few more concerns around the same logic and also some questions for @cjellick or @ibuildthecloud about the allocation framework.

**Scenario 1) ** Consider we are allocating an Instance which has multiple constraints - one of the constraints is the diskConstraint. Say other constraint is affinityConstraint. Suppose the diskConstraint passes, but say the affinityConstraint fails. Now:

Issue 1)  What should happen to the "reserved" disksize? Is it cleared out? 
Issue 2)  Do we have retries of the whole allocate() logic when one allocation try fails due to some constraint failure? @cjellick @ibuildthecloud ?
If there are retries, will the DiskConstraint that had passed previously, reserve the disk space again?

**Scenario 2)** @cjellick @ibuildthecloud  Can we have multiple Instances getting allocated simultaneously? I see that we lock at the start of by using 
lockManager.lock(new AllocateResourceLock(request), new LockCallback<Boolean>() {..}
But what does it lock on? 

I am asking because the logic above only considers the multiple volumes of the same instance that is being allocated, to calculate the "freeSize" of a given disk.
Don't we need to consider volumes of other instances that may have "reserved" space on the same disk?